common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java

```diff
@@ -87,15 +92,16 @@ public static void submitSilent(@NonNull final BaseCommand command) {
                 "Beginning execution of silent command."
         );
 
-        if (sExecutingCommands.contains(command)) {
-            new Handler(Looper.getMainLooper()).post(new Runnable() {
-                @Override
-                public void run() {
-                    command.getCallback().onError(new ClientException(ClientException.DUPLICATE_COMMAND, "The same command was already received and is being processed."));
-                }
-            });
-        } else if (command.isEligibleForCaching()) {
-            sExecutingCommands.add(command);
+        final Handler handler = new Handler(Looper.getMainLooper());
+        ResultFuture<CommandResult> future = sExecutingCommandMap.get(command);
+
+        if (null != future) {
+            future.whenComplete(getCommandResultConsumer(command, handler));
+            return;
+        } else {
+            final ResultFuture<CommandResult> resultFuture = new ResultFuture<>();
+            resultFuture.whenComplete(getCommandResultConsumer(command, handler));
+            sExecutingCommandMap.put(command, resultFuture);
```

Comment:
OK, this is your tricky synchronization part using concurrentMap.  You don't want to risk overwriting a concurrent writer without synchronizing this entire method.  So you make the result future and it's empty.  You then putIfAbsent into the sExecutingCommandMap.  If it returns null, then it did not populate the map and you add your listener to the future that you created since it was added to the map.  If it returns non-null, then you lost the race and you need add a listener to that future and return, since the command is already running.

Reply:
>If it returns null, then it did not populate the map and you add your listener to the future that you created since it was added to the map. If it returns non-null, then you lost the race and you need add a listener to that future and return, since the command is already running.

This seems backwards -- wouldn't returning `null` indicate that I _did_ win? Looking at:
https://developer.android.com/reference/java/util/concurrent/ConcurrentHashMap?hl=en#putIfAbsent(K,%20V)

Reply:
You're right.  And my statement was internally inconsistent, even.  If it returns null, it *did* populate the map and you add the callback to the future that you created.  If it returns non-null, add and return.

Reply:
Think we're saying the same thing; just read too quickly

Reply:
Ah gotcha

Reply:
> Think we're saying the same thing?

Yes.  I'm trying to agree. ;-)

Reply:
@AdamBJohnsonx Pushed changes in 98528c6c95c4eefe9d806b5f465bc8965b7f4fcb -- can you spot check me on this commit?